### PR TITLE
Support authentication in MlflowClient

### PR DESCRIPTION
**Pull request by t-henri**
## What changes are proposed in this pull request?

(Please fill in changes proposed in this fix)

## How is this patch tested?

(Details)

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/tracking/criteo_authentication.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+from mlflow.utils.rest_utils import MlflowHostCreds
+from mlflow.tracking._tracking_service.utils import (
+    _tracking_store_registry,
+    _TRACKING_TOKEN_ENV_VAR,
+)
+
+
+def get_tracking_server_uri() -> str:
+    env = os.getenv("CRITEO_ENV", "preprod")
+    return "https://mlflow.par." + env + ".crto.in"
```
Comment (jdlesage) [inline mlflow/tracking/criteo_authentication.py:18]: Could be dev. I suggest more `domain = "prod" if env == "prod" else "preprod"`

File: tests/tracking/test_criteo_authentication.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            _set_canonicalize_hostname_false(config_file)
+            with open("/tmp/krb.hadoop.jtc.conf", "r") as f:
+                krb5_conf = f.read()
+                assert "dns_canonicalize_hostname = false" in krb5_conf
+                assert "dns_canonicalize_hostname = true" not in krb5_conf
+            os.remove("/tmp/krb.hadoop.jtc.conf")
+
+
+def test_set_canonicalize_hostname_false_on_non_existing_canonicalize_hostname():
+    if sys.platform != "win32":
```
Comment (jdlesage) [inline tests/tracking/test_criteo_authentication.py:63]: Use skip_if from pytest

File: tests/tracking/test_criteo_authentication.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    mlflow.set_tracking_uri(get_tracking_server_uri())
+    mlflow.get_experiment("0")
+    assert responses.calls[0].request.headers["Authorization"] == "Bearer my-token"
+    for scheme in ["http", "https"]:
+        _tracking_store_registry.register(scheme, old_store)
+    del os.environ[_TRACKING_TOKEN_ENV_VAR]
+
+
+def test_set_canonicalize_hostname_false_on_existing_canonicalize_hostname():
+    if sys.platform != "win32":
```
Comment (jdlesage) [inline tests/tracking/test_criteo_authentication.py:49]: Use skip_if from pytest

File: mlflow/tracking/criteo_authentication.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+            + config_file
+            + " | sed 's/\\[libdefaults\\]/\\[libdefaults\\]\\n  dns_canonicalize_hostname = false/"
+            "' > /tmp/krb.hadoop.jtc.conf"
+        )
+        subprocess.check_output(cmd, shell=True)
+        os.environ["KRB5_CONFIG"] = "/tmp/krb.hadoop.jtc.conf"
+
+
+def register_criteo_authenticated_rest_store() -> None:
+    for scheme in ["http", "https"]:
```
Comment (jdlesage) [inline mlflow/tracking/criteo_authentication.py:60]: How does it work if we use it in local ? (for dev)
Is it possible to do a special case for CRITEO_ENV == dev or not exist that does not authenticate mlflow.

Authentication must be disabled if we do dev. Could you please check if it is the case with this code ?

**Review by jdlesage — CHANGES_REQUESTED** [review]

File: mlflow/tracking/criteo_authentication.py (hunk 1 above)
Comment (t-henri) [reply mlflow/tracking/criteo_authentication.py]: Done

File: tests/tracking/test_criteo_authentication.py (hunk 3 above)
Comment (t-henri) [reply tests/tracking/test_criteo_authentication.py]: Done

File: tests/tracking/test_criteo_authentication.py (hunk 2 above)
Comment (t-henri) [reply tests/tracking/test_criteo_authentication.py]: Done

File: mlflow/tracking/criteo_authentication.py (hunk 4 above)
Comment (t-henri) [reply mlflow/tracking/criteo_authentication.py:61]: Done, I handled it in client.py (call this function only if criteo_env is not dev)

**Review by jdlesage — DISMISSED** [review]

**Review by jdlesage — DISMISSED** [review]

**Review by jdlesage — APPROVED** [review]

**Pull request merged**